tunnel/src/tunnel-client.ts

```diff
@@ -121,24 +151,48 @@ export class TunnelHead extends TunnelClient {
     error: ClientErrorEvent
     response: TunnelResponse
     open: ClientOpenEvent
+    hello: {}
   }> = this._ev
 
   public constructor(public readonly tunnelId: string, ws: WebSocket) {
     super(ws)
 
     this._ev.on('message', (ev: WebSocket.MessageEvent) => {
-      const strData = ev.data.toString()
-      const parseResult = tunnelResponseSchema.safeParse(JSON.parse(strData))
-      if (!parseResult.success) {
+      const message = this._parseMessage(ev)
+      if (!message) {
         this.close(errors.CLOSE_CODES.INVALID_RESPONSE_PAYLOAD, 'invalid response payload')
         return
       }
-      this.events.emit('response', parseResult.data)
+      if (message.type === 'hello') {
+        this.events.emit('hello', {})
+        return
+      }
+      this.events.emit('response', message.reponse)
     })
   }
 
-  public readonly send = (request: TunnelRequest) => {
+  public readonly send = (request: Omit<TunnelRequest, 'type'>) => {
     this._throwIfClosed()
-    this._ws.send(JSON.stringify(request))
+
+    const req: TunnelRequest = { type: 'request', ...request }
+    this._ws.send(JSON.stringify(req))
+  }
+
+  private _parseMessage = (
+    ev: WebSocket.MessageEvent
+  ): { type: 'hello' } | { type: 'response'; reponse: TunnelResponse } | undefined => {
+    const data = JSON.parse(ev.data.toString())
+
+    const responseParseResult = tunnelResponseSchema.safeParse(data)
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**🎯 suggestion:** ‏ You could do a discriminated union instead of parsing 2 times the same data.